## Завдання

В якості домашнього завдання вам пропонується створити нейронну мережу за допомогою механізмів Keras, яка буде класифікувати товари із датасету fasion_mnist.



Вам належить запропонувати свою власну архітектуру мережі. Точність найнаївнішої, але адекватної нейромережі становить приблизно 91%. Точність вашої моделі повинна бути не нижчою за цей показник. Щоб досягти таких значень вам знадобиться поекспериментувати з гіперпараметрами мережі:



кількість шарів;
кількість нейронів;
функції активації;
кількість епох;
розмір батчу;
вибір оптимізатора;
різні техніки регуляризації і т.д.


Використайте вивчені техніки виявлення проблем навчання нейронної мережі, і потім поекспериментуйте.



Рішення оформіть у вигляді окремого ноутбука.

## Імпорт необхідних бібліотек

In [17]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten, Conv2D, MaxPooling2D, Dropout
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.regularizers import l2
from tensorflow.keras.optimizers import RMSprop, Nadam


## Завантаження і підготовка даних

In [2]:
# Завантаження даних
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

# Нормалізація даних
x_train = x_train / 255.0
x_test = x_test / 255.0

# Зміна форми даних
x_train = x_train.reshape(-1, 28, 28, 1)
x_test = x_test.reshape(-1, 28, 28, 1)

# One-hot encoding міток
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)


4422102/4422102 [==============================] - 1s 0us/step


## Архітектура моделі

In [18]:
# Створення моделі
model = Sequential()

# Перший згортковий шар
model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1), kernel_regularizer=l2(0.001)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

# Другий згортковий шар
model.add(Conv2D(64, kernel_size=(3, 3), activation='relu', kernel_regularizer=l2(0.001)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

# Третій згортковий шар
model.add(Conv2D(128, kernel_size=(3, 3), activation='relu', kernel_regularizer=l2(0.001)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

# Повнозв'язний шар
model.add(Flatten())
model.add(Dense(256, activation='relu', kernel_regularizer=l2(0.001)))
model.add(Dropout(0.5))
model.add(Dense(128, activation='relu', kernel_regularizer=l2(0.001)))
model.add(Dropout(0.5))

# Вихідний шар
model.add(Dense(10, activation='softmax'))

# Компіляція моделі
#model.compile(optimizer=Adam(), loss='categorical_crossentropy', metrics=['accuracy'])
model.compile(optimizer=RMSprop(), loss='categorical_crossentropy', metrics=['accuracy'])


## Тренування моделі

In [19]:
# Тренування моделі
model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=30, batch_size=128)


Epoch 1/30
469/469 [==============================] - 61s 129ms/step - loss: 1.3272 - accuracy: 0.6083 - val_loss: 0.8535 - val_accuracy: 0.7686
Epoch 2/30
469/469 [==============================] - 57s 122ms/step - loss: 0.8465 - accuracy: 0.7656 - val_loss: 0.7138 - val_accuracy: 0.7954
Epoch 3/30
469/469 [==============================] - 57s 122ms/step - loss: 0.7448 - accuracy: 0.7924 - val_loss: 0.6380 - val_accuracy: 0.8238
Epoch 4/30
469/469 [==============================] - 58s 125ms/step - loss: 0.6903 - accuracy: 0.8082 - val_loss: 0.6318 - val_accuracy: 0.8205
Epoch 5/30
469/469 [==============================] - 60s 127ms/step - loss: 0.6622 - accuracy: 0.8178 - val_loss: 0.5869 - val_accuracy: 0.8331
Epoch 6/30
469/469 [==============================] - 57s 121ms/step - loss: 0.6370 - accuracy: 0.8267 - val_loss: 0.5551 - val_accuracy: 0.8529
Epoch 7/30
469/469 [==============================] - 58s 123ms/step - loss: 0.6210 - accuracy: 0.8317 - val_loss: 0.5410 - val_ac

KeyboardInterrupt: 

## Оцінка моделі

In [14]:
# Оцінка моделі
score = model.evaluate(x_test, y_test, verbose=0)
print(f'Test loss: {score[0]}')
print(f'Test accuracy: {score[1]}')


Test loss: 0.4837980568408966
Test accuracy: 0.8828999996185303


## Експерименти

# 0.
Test loss: 0.3074522316455841
Test accuracy: 0.8870000243186951

#1.
Проведемо збільшення кількості епох з 15 до 30.
Було:
```
model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=15, batch_size=128)
```
Стало:
```
model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=30, batch_size=128)
```
Test loss: 0.26392775774002075
Test accuracy: 0.8996999859809875


# 2.
Зміна архітектури моделі.
Додамо ще один повнозв'язний шар і збільшимо кількість нейронів.
Було:
```
# Повнозв'язний шар
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
```

Стало:
```
model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
```

Тест:
Test loss: 0.2852315604686737
Test accuracy: 0.8934000134468079

3. Додавання L2-регуляризації

Стало:

```
from tensorflow.keras.regularizers import l2

model = Sequential()

model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1), kernel_regularizer=l2(0.001)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Conv2D(64, kernel_size=(3, 3), activation='relu', kernel_regularizer=l2(0.001)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Conv2D(128, kernel_size=(3, 3), activation='relu', kernel_regularizer=l2(0.001)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Flatten())
model.add(Dense(256, activation='relu', kernel_regularizer=l2(0.001)))
model.add(Dropout(0.5))
model.add(Dense(128, activation='relu', kernel_regularizer=l2(0.001)))
model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

model.compile(optimizer=Adam(), loss='categorical_crossentropy', metrics=['accuracy'])

```

Test loss: 0.4837980568408966
Test accuracy: 0.8828999996185303

4. Зміна оптимізатора.

RMSprop

```
# Використання RMSprop
from tensorflow.keras.optimizers import RMSprop
model.compile(optimizer=RMSprop(), loss='categorical_crossentropy', metrics=['accuracy'])
```

Test RMSprop:

Nadam
```
# Використання Nadam
model.compile(optimizer=Nadam(), loss='categorical_crossentropy', metrics=['accuracy'])
```